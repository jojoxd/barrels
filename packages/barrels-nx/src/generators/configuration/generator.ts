import { formatFiles, generateFiles, Tree } from '@nx/devkit';
import { join as joinPath } from 'path';
import { Schema } from './schema';
import {addBarrelTask} from "./lib/util-functions";

export async function configurationGenerator(tree: Tree, options: Schema): Promise<void>
{
  const projectRoot = `libs/${options.name}`;

  generateFiles(tree, joinPath(__dirname, 'files'), projectRoot, options);

  await addBarrelTask(tree, options);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default configurationGenerator;
