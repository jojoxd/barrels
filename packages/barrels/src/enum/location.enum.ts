export enum Location
{
    /**
     * Create a single barrel in the top-most directory
     */
    Top = 'top',

    /**
     * Create a barrel in all directories that contain files, excluding the top directory
     */
    Branch = 'branch',

    /**
     * Create a barrel in all directories that contain files, including the top directory
     */
    All = 'all',
}
