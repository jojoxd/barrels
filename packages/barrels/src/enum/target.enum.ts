export enum Target
{
    /**
     * Generate typescript
     */
    Typescript = 'ts',

    /**
     * Generate .d.ts files only
     */
    TypescriptTypes = 'dts',

    /**
     * Generate commonjs
     */
    CommonJS = 'cjs',

    /**
     * generate modules
     */
    Module = 'esm',
}
