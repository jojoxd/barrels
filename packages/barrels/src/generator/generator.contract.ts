import {BarrelContext} from "../model/barrel-context.js";

export interface GeneratorContract
{
    supports(context: BarrelContext): boolean;

    createExportStatement(context: BarrelContext): Promise<string>;
}
