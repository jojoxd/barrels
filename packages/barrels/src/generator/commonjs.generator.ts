import {GeneratorContract} from "./generator.contract.js";
import {BarrelContext} from "../model/barrel-context.js";
import {Target} from "../enum/target.enum.js";

export class CommonJSGenerator implements GeneratorContract
{
    public supports(context: BarrelContext): boolean
    {
        return [
            Target.CommonJS,
        ].includes(context.configuration.target);
    }

    public async createExportStatement(context: BarrelContext): Promise<string>
    {
        throw new Error("Generating CommonJS is not supported yet");
    }
}