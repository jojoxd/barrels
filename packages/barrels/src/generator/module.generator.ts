import {GeneratorContract} from "./generator.contract.js";
import {BarrelContext} from "../model/barrel-context.js";
import {ast, AstTransformer} from "../service/ast-transformer.js";
import {Target} from "../enum/target.enum.js";
import {AssertClauseFactory} from "./ast-factory/assert-clause.factory.js";

export class ModuleGenerator implements GeneratorContract
{
    public supports(context: BarrelContext): boolean
    {
        return [
            Target.Module,
            Target.Typescript,
            Target.TypescriptTypes,
        ].includes(context.configuration.target);
    }

    public async createExportStatement(context: BarrelContext): Promise<string>
    {
        const file = 'todo.ts';

        // @TODO: Generate ast

        const moduleSpecifier = ast.createStringLiteral(file);

        const assertClause = context.emitAssertions ? AssertClauseFactory.createAssertClause(file, context) : undefined;

        const exportDeclaration = ast.createExportDeclaration(
            undefined,
            context.configuration.target === Target.TypescriptTypes,
            undefined,
            moduleSpecifier,
            assertClause,
        );

        const nodeArray = ast.createNodeArray([exportDeclaration]);

        const transformer = new AstTransformer(context);
        return transformer.transform(nodeArray);
    }
}
