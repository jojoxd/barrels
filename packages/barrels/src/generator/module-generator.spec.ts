import {ModuleGenerator} from "./module.generator.js";
import {Target} from "../enum/target.enum.js";
import { createBarrelsConfiguration } from '../../tests/support/create-barrels-configuration.js';

const moduleGenerator = new ModuleGenerator();

const supportedTargets = [
    Target.Module,
    Target.Typescript,
    Target.TypescriptTypes
];

it.each(supportedTargets)('Supports Target "%s"', async (target: Target) => {
    const { context } = await createBarrelsConfiguration({
        target,
        barrels: [{ directory: 'empty', }],
    });

    expect(moduleGenerator.supports(context)).toBeTruthy();
});

describe('Generation', () => {
    it('Should generate assertions when enabled', async () => {
        const { context } = await createBarrelsConfiguration({
            target: Target.Module,
            barrels: [{ directory: 'empty' }],
            emitAssertions: true,
        });

        const exportStatement = await moduleGenerator.createExportStatement(context);
        expect(exportStatement).toMatchSnapshot();
    });

    it('Should not generate assertions when disabled', async () => {
        const { context } = await createBarrelsConfiguration({
            target: Target.Module,
            barrels: [{ directory: 'empty' }],
            emitAssertions: false,
        });

        const exportStatement = await moduleGenerator.createExportStatement(context);
        expect(exportStatement).toMatchSnapshot();
    });
});
