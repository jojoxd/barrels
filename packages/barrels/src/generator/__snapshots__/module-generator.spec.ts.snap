// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generation Should generate assertions when enabled 1`] = `
"    export * from "todo.ts" assert { "type": "json" };
"
`;

exports[`Generation Should not generate assertions when disabled 1`] = `
"    export * from "todo.ts";
"
`;
